pkg/detectors/npm/registry.go

```diff
@@ -0,0 +1,300 @@
+		// https://help.sonatype.com/repomanager2/node-packaged-modules-and-npm-registries
+		sb.WriteString(`|(` + hostPat + `/nexus/content/(?:groups|repositories)/[a-z0-9-][a-z0-9._-]+)`)
+		// https://help.sonatype.com/repomanager3/nexus-repository-administration/formats/npm-registry/configuring-npm
+		sb.WriteString(`|(` + hostPat + `/(?:nexus/)?repository/[a-z0-9-][a-z0-9._-]+)`)
+		// https://docs.gitlab.com/ee/user/packages/npm_registry/
+		sb.WriteString(`|(` + hostPat + `/api/v4/(?:groups/\d+/-/|projects/\d+/)?packages/npm)`)
+		// https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry
+		sb.WriteString(`|(npm\.pkg\.github\.com)`)
+		// https://learn.microsoft.com/en-us/azure/devops/artifacts/get-started-npm?view=azure-devops&tabs=Windows
+		// https://stackoverflow.com/a/73495381
+		azurePat := `pkgs\.dev\.azure\.com/[a-z0-9._-]+(?:/[a-z0-9._-]+)?`
+		vsPat := `[a-z0-9-]+\.pkgs\.visualstudio\.com(?:/[a-z0-9._-]+)?`
+		sb.WriteString(fmt.Sprintf(`|((?:%s|%s)/_packaging/[a-z0-9._-]+/npm(?:/registry)?)`, azurePat, vsPat))
+		// https://www.jetbrains.com/help/space/using-an-npm-registry-with-npmjs-com.html
+		sb.WriteString(`|(npm\.pkg\.jetbrains\.space/[a-z0-9._-]+/p/[a-z0-9._-]+/[a-z0-9._-]+)`)
+		sb.WriteString(`|((?:[a-z0-9-]+-)?npm\.pkg\.dev/[a-z0-9._-]+/[a-z0-9._-]+)`)
+		sb.WriteString(`|(npm(?:-proxy)?\.fury\.io/[a-z0-9._-]+)`)
+		sb.WriteString(`)`)
+
+		return regexp.MustCompile(sb.String())
+	}()
+	genericRegistryPat = func() *regexp.Regexp {
+		urlPat := fmt.Sprintf(`%s(?:/[a-z0-9._-]+)*`, hostPat)
+		registryPat := regexp.MustCompile(fmt.Sprintf(
+			`(?i)['"]?(//%s)/:|registry.{1,50}?['"]?(https?://%s)/?['"]?|@[a-z0-9\-_]{1,50}['"]?[ \t]*(?:=[ \t]*)?['"]?(https?://%s)/?['"]?|\[npmAuth\.['"](https?://%s)/?['"]\]`, urlPat, urlPat, urlPat, urlPat))
+
+		// Sanity check to make sure the pattern doesn't contain a mistake.
+		if registryPat.NumSubexp() != 4 {
+			panic(fmt.Sprintf("Pattern |genericRegistryPat| should have 4 capture groups but has %d", registryPat.NumSubexp()))
+		}
+		return registryPat
+	}()
+)
+
+// findTokenRegistry returns the specific registry associated with the |token| if a high confidence match is found in |data|.
+//
+// Common configurations:
+// - npm: https://docs.npmjs.com/using-private-packages-in-a-ci-cd-workflow#create-and-check-in-a-project-specific-npmrc-file
+// - Yarn (TODO)
+// - Unity Package Manager (TODO)
+func findTokenRegistry(data string, token string) *registryInfo {
+	// .npmrc stores auth as `//registry.com/path/:authToken=$TOKEN
+	// Therefore, we should be able to correlate a token to a registry with a high degree of confidence.
+	registryAuthPat := regexp.MustCompile(fmt.Sprintf(
+		// language=regexp
+		`(?i)(//%s(?:/[a-z0-9._-]+)*)/:(?:_auth(?:Token)?|_password).{1,20}%s`, hostPat, token))
+	matches := registryAuthPat.FindStringSubmatch(data)
+	if len(matches) == 0 {
```

Comment:
I don't believe so. If a regex pattern matches, it always returns a fixed number of match groups. If you have mutually exclusive match groups like `(?:(foo)|(bar))` then the non-matching group will just be an empty string.


```go
func main() {
	pat := regexp.MustCompile(`(?i)hello (?:(john)|(anne)|(world)|(.+))`)
	matches := pat.FindStringSubmatch("Hello Anne")
	fmt.Printf("Matches return %d groups: %s\n", len(matches), strings.Join(matches, ", "))
}

// Matches return 5 groups: Hello Anne, , Anne, , 
```
https://go.dev/play/p/cdPwcqZwDcT

Reply:
I suppose you could make the argument that it prevents failure, but if your pattern changes I think it's better to explicitly fail than silently break.

e.g., stuff like this will silently break if the pattern is updated to add another match group.
https://github.com/trufflesecurity/trufflehog/blob/5e78ad402e7f98817b18a93e4e5807ada53bad4d/pkg/detectors/aha/aha.go#L40-L42